ice-adapter/src/main/java/com/faforever/iceadapter/ice/Peer.java

```diff
@@ -38,26 +44,32 @@ public Peer(GameSession gameSession, int remoteId, String remoteLogin, boolean l
                 String.valueOf(localOffer),
                 preferredPort);
 
-        initForwarding(preferredPort);
+        faSocket = initForwarding(preferredPort);
+
+        AsyncService.runAsync(this::faListener);
 
         if (localOffer) {
-            new Thread(ice::initiateIce).start();
+            AsyncService.runAsync(ice::initiateIce);
         }
     }
 
+    public int getLocalPort() {
+        return faSocket.getLocalPort();
+    }
+
     /**
      * Starts waiting for data from FA
      */
-    private void initForwarding(int port) {
+    @SneakyThrows(SocketException.class)
+    private DatagramSocket initForwarding(int port) {
         try {
-            faSocket = new DatagramSocket(port);
+            DatagramSocket socket = new DatagramSocket(port);
+            log.debug("Now forwarding data to peer {}", getPeerIdentifier());
+            return socket;
         } catch (SocketException e) {
             log.error("Could not create socket for peer: {}", getPeerIdentifier(), e);
+            throw e;
```

Comment:
If the socket has returned an error, then it doesn't make sense further. Because we are going to get a LocalPort, and the socket will be null.